Build failure: kotatogram-desktop -> tg_owt

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

With the most recent packages (`tg_owt-0-unstable-2024-06-15`), a rebuild results in build errors.

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/325046031

### Relevant log output

```console
[266/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/nv12_buffer.cc.o
FAILED: [code=1] CMakeFiles/tg_owt.dir/src/api/video/nv12_buffer.cc.o 
/nix/store/hb2bs5fg5wkm04x565737qd5nh2hy5nk-gcc-wrapper-15.2.0/bin/g++ -DABSL_ALLOCATOR_NOTHROW=1 -DBWE_TEST_LOGGING_COMPILE_TIME_ENABLE=0 -DHAVE_NETINET_IN_H -DHAVE_SCTP -DHAVE_WEBRTC_VIDEO -DNDEBUG -DNO_MAIN_THREAD_WRAPPING -DRTC_DISABLE_TRACE_EVENTS -DRTC_ENABLE_VP9 -DWEBRTC_APM_DEBUG_DUMP=0 -DWEBRTC_DUMMY_AUDIO_BUILD -DWEBRTC_ENABLE_PROTOBUF=0 -DWEBRTC_HAVE_DCSCTP -DWEBRTC_HAVE_SCTP -DWEBRTC_INCLUDE_INTERNAL_AUDIO_DEVICE -DWEBRTC_LIBRARY_IMPL -DWEBRTC_LINUX -DWEBRTC_NON_STATIC_TRACE_EVENT_HANDLERS=1 -DWEBRTC_OPUS_SUPPORT_120MS_PTIME=1 -DWEBRTC_OPUS_VARIABLE_COMPLEXITY=0 -DWEBRTC_POSIX -DWEBRTC_USE_BUILTIN_ISAC_FLOAT -DWEBRTC_USE_H264 -DWEBRTC_USE_PIPEWIRE -DWEBRTC_USE_X11 -I/build/source/src -I/build/source/src/third_party/pffft/src -I/build/source/src/third_party/libsrtp/include -I/build/source/src/third_party/libsrtp/crypto/include -I/build/source/src/third_party/libsrtp/../libsrtp_config -I/build/source/src/third_party/libyuv/include -isystem /nix/store/4lsi1w9paqi22y1vn29ix5l5gvjikzqy-glib-2.86.3-dev/include/gio-unix-2.0 -isystem /nix/store/4lsi1w9paqi22y1vn29ix5l5gvjikzqy-glib-2.86.3-dev/include/glib-2.0 -isystem /nix/store/d864rlg855kf2hnnnfdzcx54grp8n10x-glib-2.86.3/lib/glib-2.0/include -isystem /nix/store/ha1xraqy2kpbv48n709nxxislzinyq01-pipewire-1.6.2-dev/include/pipewire-0.3 -isystem /nix/store/ha1xraqy2kpbv48n709nxxislzinyq01-pipewire-1.6.2-dev/include/spa-0.2 -isystem /nix/store/nm8vpqbxi7qx1dgq5nb4arj0hqhbxvby-libdrm-2.4.131-dev/include/libdrm -isystem /nix/store/zpyr122jwgf1d6irwabv2b5788dhm1jz-libopus-1.6.1-dev/include/opus -O3 -DNDEBUG -std=gnu++20 -Wno-deprecated-declarations -Wno-attributes -Wno-narrowing -Wno-return-type -MD -MT CMakeFiles/tg_owt.dir/src/api/video/nv12_buffer.cc.o -MF CMakeFiles/tg_owt.dir/src/api/video/nv12_buffer.cc.o.d -o CMakeFiles/tg_owt.dir/src/api/video/nv12_buffer.cc.o -c /build/source/src/api/video/nv12_buffer.cc
/build/source/src/api/video/nv12_buffer.cc: In member function 'void webrtc::NV12Buffer::InitializeData()':
/build/source/src/api/video/nv12_buffer.cc:122:3: error: 'memset' was not declared in this scope
  122 |   memset(data_.get(), 0, NV12DataSize(height_, stride_y_, stride_uv_));
      |   ^~~~~~
/build/source/src/api/video/nv12_buffer.cc:18:1: note: 'memset' is defined in header '<cstring>'; this is probably fixable by adding '#include <cstring>'
   17 | #include "third_party/libyuv/include/libyuv/scale.h"
  +++ |+#include <cstring>
   18 | 
[267/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/i444_buffer.cc.o
[268/1394] Building CXX object CMakeFiles/tg_owt.dir/src/call/adaptation/resource_adaptation_processor_interface.cc.o
[269/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/units/data_rate.cc.o
[270/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/video_bitrate_allocation.cc.o
[271/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/i210_buffer.cc.o
[272/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/wrapping_async_dns_resolver.cc.o
[273/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/transport/rtp/dependency_descriptor.cc.o
[274/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/video_adaptation_counters.cc.o
[275/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/field_trials.cc.o
[276/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/candidate.cc.o
[277/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/color_space.cc.o
[278/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/i420_buffer.cc.o
[279/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/video_bitrate_allocator.cc.o
[280/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/units/time_delta.cc.o
[281/1394] Building CXX object CMakeFiles/tg_owt.dir/src/call/adaptation/resource_adaptation_processor.cc.o
[282/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/units/frequency.cc.o
[283/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/video_frame_buffer.cc.o
[284/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/stats_types.cc.o
[285/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/units/data_size.cc.o
[286/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/video_frame.cc.o
[287/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/transport/network_types.cc.o
[288/1394] Building CXX object CMakeFiles/tg_owt.dir/src/call/adaptation/video_stream_input_state_provider.cc.o
[289/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/encoded_frame.cc.o
[290/1394] Building CXX object CMakeFiles/tg_owt.dir/src/audio/voip/audio_egress.cc.o
[291/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/builtin_video_bitrate_allocator_factory.cc.o
[292/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/encoded_image.cc.o
[293/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/create_peerconnection_factory.cc.o
[294/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/transport/stun.cc.o
[295/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/rtp_parameters.cc.o
[296/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/peer_connection_interface.cc.o
[297/1394] Building CXX object CMakeFiles/tg_owt.dir/src/api/video/rtp_video_frame_assembler.cc.o
```

### Additional context

_No response_

### System metadata

```logs
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.19.8-cachyos, NixOS, 26.05 (Yarara), 26.05.20260318.b40629e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.3`
 - nixpkgs: `/nix/store/ap9dpkyzikzzh04259wlsvha2mw455x4-source`
```

### Notify maintainers

@ilya-fedin 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/cstring-includes.patch
+++ b/pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/cstring-includes.patch
@@ -1,0 +1,74 @@
+diff --git a/src/api/video/nv12_buffer.cc b/src/api/video/nv12_buffer.cc
+index ca9dcd867..89d28f23c 100644
+--- a/src/api/video/nv12_buffer.cc
++++ b/src/api/video/nv12_buffer.cc
+@@ -16,6 +16,8 @@
+ #include "third_party/libyuv/include/libyuv/convert.h"
+ #include "third_party/libyuv/include/libyuv/scale.h"
+
++#include <cstring>
++
+ namespace webrtc {
+
+ namespace {
+diff --git a/src/audio/utility/channel_mixer.cc b/src/audio/utility/channel_mixer.cc
+index 0f1e66387..33b771b0c 100644
+--- a/src/audio/utility/channel_mixer.cc
++++ b/src/audio/utility/channel_mixer.cc
+@@ -15,6 +15,8 @@
+ #include "rtc_base/logging.h"
+ #include "rtc_base/numerics/safe_conversions.h"
+
++#include <cstring>
++
+ namespace webrtc {
+
+ ChannelMixer::ChannelMixer(ChannelLayout input_layout,
+diff --git a/src/modules/audio_processing/aec3/alignment_mixer.cc b/src/modules/audio_processing/aec3/alignment_mixer.cc
+index 7f076dea8..ffd7242b5 100644
+--- a/src/modules/audio_processing/aec3/alignment_mixer.cc
++++ b/src/modules/audio_processing/aec3/alignment_mixer.cc
+@@ -10,6 +10,7 @@
+ #include "modules/audio_processing/aec3/alignment_mixer.h"
+
+ #include <algorithm>
++#include <cstring>
+
+ #include "rtc_base/checks.h"
+
+diff --git a/src/modules/desktop_capture/linux/wayland/shared_screencast_stream.cc b/src/modules/desktop_capture/linux/wayland/shared_screencast_stream.cc
+index 7ef1a030e..5b9ab7137 100644
+--- a/src/modules/desktop_capture/linux/wayland/shared_screencast_stream.cc
++++ b/src/modules/desktop_capture/linux/wayland/shared_screencast_stream.cc
+@@ -18,6 +18,7 @@
+ #include <spa/param/video/format-utils.h>
+ #include <sys/mman.h>
+
++#include <cstring>
+ #include <vector>
+
+ #include "absl/memory/memory.h"
+diff --git a/src/modules/video_coding/utility/ivf_file_reader.cc b/src/modules/video_coding/utility/ivf_file_reader.cc
+index 4c08ca613..f82f2bfcb 100644
+--- a/src/modules/video_coding/utility/ivf_file_reader.cc
++++ b/src/modules/video_coding/utility/ivf_file_reader.cc
+@@ -10,6 +10,7 @@
+
+ #include "modules/video_coding/utility/ivf_file_reader.h"
+
++#include <cstring>
+ #include <string>
+ #include <vector>
+
+diff --git a/src/net/dcsctp/packet/bounded_byte_writer.h b/src/net/dcsctp/packet/bounded_byte_writer.h
+index d754549e4..bf5e3ed42 100644
+--- a/src/net/dcsctp/packet/bounded_byte_writer.h
++++ b/src/net/dcsctp/packet/bounded_byte_writer.h
+@@ -12,6 +12,7 @@
+ #define NET_DCSCTP_PACKET_BOUNDED_BYTE_WRITER_H_
+
+ #include <algorithm>
++#include <cstring>
+
+ #include "api/array_view.h"
+ 

--- a/pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
+++ b/pkgs/applications/networking/instant-messengers/telegram/kotatogram-desktop/default.nix
@@ -31,6 +31,9 @@ let
     };
 
     patches = [
+      # fix build with latest glibc
+      # upstream PR: https://github.com/desktop-app/tg_owt/pull/172
+      ./cstring-includes.patch
       (fetchpatch {
         url = "https://webrtc.googlesource.com/src/+/e7d10047096880feb5e9846375f2da54aef91202%5E%21/?format=TEXT";
         decode = "base64 -d";

